### ISSUE TITLE
secrets: service update --secret-rm removes by "source", not "target"

### ISSUE THREAD

**Issue by thaJeztah** [docker/docker#28956]
Doing more test on consistency, it looks like `--secret-rm` is using the wrong property for removing a secret;

To reproduce; create a secret, and a service that uses the same secret _twice_:

```bash
echo "world" | docker secret create hello

docker service create --name test --secret source=hello,target=world --secret source=hello,target=other-world nginx:alpine
```

Verify that the secret is used twice;

```bash
docker service inspect test --format '{{ json .Spec.TaskTemplate.ContainerSpec.Secrets}}' | jq
```

```json
[
  {
    "File": {
      "Name": "world",
      "UID": "0",
      "GID": "0",
      "Mode": 292
    },
    "SecretID": "jis2m9o0w5b2h4b1hu1ejjbwa",
    "SecretName": "hello"
  },
  {
    "File": {
      "Name": "other-world",
      "UID": "0",
      "GID": "0",
      "Mode": 292
    },
    "SecretID": "jis2m9o0w5b2h4b1hu1ejjbwa",
    "SecretName": "hello"
  }
]
```

Update the service, and remove a secret;

```bash
docker service update --secret-rm hello test
```

Both uses of the secret are removed;

```bash
docker service inspect test --format '{{ json .Spec.TaskTemplate.ContainerSpec.Secrets}}' | jq
null
```

The _target_ should be leading here, because a single secret can be used multiple times. This is also consistent with other options (e.g. `--mount-rm`)

/cc @stevvooe @ehazlett

**Comment by aluzzardi** [comment]
P1'ing this one since it's UX and we have to figure this out before the release

**Comment by diogomonica** [comment]
This works as intended, no change is needed.

**Comment by thaJeztah** [comment]
But the _target_ is the canonical here. Similar to `--mount`;

Create a service, using the same volume twice (different mount-points);

```bash
docker volume create hello
docker service create --name test --mount source=hello,target=world --mount source=hello,target=other-world nginx:alpine
```

Removing the mount is done by _target_;

```bash
docker service update --mount-rm other-world test
```

**Comment by diogomonica** [comment]
Actually, the target is canonical per service, but the name is canonical per-swarm.

Anyway, not entirely sure I care enough to fight this normalization. I'm fine with whichever.

**Comment by stevvooe** [comment]
Unfortunately, I think _everyone_ is right here.

Let's break this down. First of all, we have to take into account that we aren't actually operating on secrets, we are operating on _secret references_. If we look at this way, we can indeed say they are somewhat equivalent to mounts, though we need to be careful extending that analogy too far.

In general, the user needs to be able to do the following:

1. Remove a single reference to a secret, keyed by the target.
2. Remove _all_ references to a secret, by secret name (or id, theoretically).

If we model each reference as a tuple, `(<name>, <target>)` we can analyze this more easily. Essentially, we are looking for a model that allows deterministic removal of a _secret reference_ through only the `--secret-rm` flag. In the case presented by @thaJeztah, we have the following:

```
(hello, world)
(hello, other-world)
```

With the current behavior, we can issue `--secret-rm hello`, `--secret-rm world` and `--secret-rm other-world` and get deterministic behavior, per our requirements above. Respectively, we remove all references to `hello`, or just the reference to `world` and `other-world`.

The problem with this model when we have something more ambiguous. Let's extend @thaJeztah's case with more data:

```
(hello, world)
(hello, other-world)
(world, foo)
```

Here, if we issue `--secret-rm world`, do we remove the references to the `world` secret or the secret references with the target as `world`, or both? Typically, the key to such situations is to detect the ambiguity and provide user feedback. This might look like the following:

```
$ docker service update --secret-rm world foo
Error: ambiguos secret "world", please specify whether to remove by source or target.
Found source=world,target=foo and source=hello,target=world
```

With the above, the user can then issue the following to remove the secret references unambiguously:

```
$ docker service update --secret-rm target=world foo
```

**Comment by diogomonica** [comment]
I agree with removing the ambiguity. Maybe for now we simply reject using the same source multiple times? source also becomes unique within a service reference.

**Comment by stevvooe** [comment]
> Maybe for now we simply reject using the same source multiple times? source also becomes unique within a service reference.

In such a scenario, do we remove on the source key or target key? I would think the target key, since that falls back to source. This case may cause a problem with the following:

```
(hello, world)
(world, world)
```

This would require source key-based removal, which is probably okay for this use case, but may be limiting as we extend this syntax.

If we want to reinforce the reference model, this might be problematic. However, for this particular use case, that really only configures file projections of secret references, this limitation is sufficient. I think we may want to support these through template references, but that can be automated without any kind of cli flag (template parsing, ftw!).

**Comment by diogomonica** [comment]
@stevvooe what is your concrete proposal moving forward then? Should we do the disambiguation changes now?

**Comment by stevvooe** [comment]
The most future-proof is the proposal from https://github.com/docker/docker/issues/28956#issuecomment-263752908. If that is out of the question, go with `source` matching. It is not as future-proof, but solves the issue. Based on @thaJeztah's description, this might already be the behavior now.

**Comment by robwithhair** [comment]
Since `--secret-add` uses `source=secret,target=name` notation would it not be wise to have similar notation in the `--secret-rm` command?  So you can either run `--secret-rm source` or `--secret-rm target=name` or even `--secret-rm source=secret,target=name` or `--secret-rm source=secret` if you wanted to.  IMO this would be simplest from a user perspective.  

I just ran into this today and have no way of removing the secret because I am only tracking it's mounted name ATM.
